org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/markers/ShowIssueFlowsMarkerResolver.java

```diff
@@ -19,77 +19,136 @@
  */
 package org.sonarlint.eclipse.ui.internal.markers;
 
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
 import org.eclipse.core.resources.IMarker;
 import org.eclipse.jface.text.BadPositionCategoryException;
 import org.eclipse.jface.text.IDocument;
 import org.eclipse.jface.text.Position;
 import org.eclipse.jface.text.source.Annotation;
 import org.eclipse.jface.text.source.IAnnotationModel;
+import org.eclipse.jface.text.source.IAnnotationModelExtension;
 import org.eclipse.swt.graphics.Image;
 import org.eclipse.ui.IEditorInput;
 import org.eclipse.ui.IEditorPart;
+import org.eclipse.ui.IEditorReference;
 import org.eclipse.ui.IMarkerResolution2;
 import org.eclipse.ui.IWorkbenchPage;
+import org.eclipse.ui.PartInitException;
 import org.eclipse.ui.PlatformUI;
 import org.eclipse.ui.ide.IDE;
+import org.eclipse.ui.texteditor.IDocumentProvider;
 import org.eclipse.ui.texteditor.ITextEditor;
 import org.sonarlint.eclipse.core.SonarLintLogger;
 import org.sonarlint.eclipse.core.internal.markers.MarkerUtils;
 import org.sonarlint.eclipse.core.internal.markers.MarkerUtils.ExtraPosition;
 import org.sonarlint.eclipse.ui.internal.SonarLintImages;
+import org.sonarlint.eclipse.ui.internal.views.IssueLocationsView;
 
 public class ShowIssueFlowsMarkerResolver implements IMarkerResolution2 {
 
   private static final String ISSUE_FLOW_ANNOTATION_TYPE = "org.sonarlint.eclipse.issueFlowAnnotationType";
 
   @Override
   public String getDescription() {
-    return "Show the extra locations to better understand the issue";
+    return "Show/Hide all locations to better understand the issue";
   }
 
   @Override
   public String getLabel() {
-    return "Show issue extra locations";
+    return "Toggle all issue locations";
   }
 
   @Override
   public void run(IMarker marker) {
     try {
       IWorkbenchPage page = PlatformUI.getWorkbench().getActiveWorkbenchWindow().getActivePage();
       IEditorPart editorPart = IDE.openEditor(page, marker);
-      IEditorInput editorInput = editorPart.getEditorInput();
       if (editorPart instanceof ITextEditor) {
         ITextEditor textEditor = (ITextEditor) editorPart;
-        IAnnotationModel annotationModel = textEditor.getDocumentProvider().getAnnotationModel(editorInput);
-        removePreviousAnnotations(annotationModel);
-        IDocument doc = textEditor.getDocumentProvider().getDocument(editorInput);
-        createAnnotations(marker, annotationModel, doc);
+        toggleAnnotations(marker, textEditor, false);
+        try {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Extract this nested try block into a separate method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1141)